### DISCUSSION TITLE
Is it possible to export constants in .NET

### DISCUSSION THREAD

**Discussion by 2A5F** [dotnet/runtime#89236] [General]
DirectX 12 Agility Sdk needs to export constants in exe, seems like this is not possible for .NET

https://devblogs.microsoft.com/directx/gettingstarted-dx12agility/#2-set-agility-sdk-parameters

> ### Method a: __declspec(dllexport) keyword
> A can export these constants as follows using the __declspec(dllexport) keyword:
> ```cpp
> extern "C" { __declspec(dllexport) extern const UINT D3D12SDKVersion = n;}
> 
> extern "C" { __declspec(dllexport) extern const char* D3D12SDKPath = u8".\\D3D12\\"; }
> ```
> Here n corresponds to the D3D12SDKVersion you’re using. For our first Agility SDK, this value is 4.
> 
> ### Method b: Module-definition file
> Another option is to export these constants via a .def file:
> ```
> EXPORTS
> 
>   D3D12SDKVersion DATA PRIVATE
> 
>   D3D12SDKPath DATA PRIVATE
> ```
> And then to declare the constants in code as follows:
> ```cpp
> extern "C" extern const uint32_t D3D12SDKVersion = n;
> 
> extern "C" extern LPCSTR D3D12SDKPath = ".\\D3D12\\";
> ```
> Here n corresponds to the D3D12SDKVersion you’re using. For our first Agility SDK, this value is 4.

**Comment by DaZombieKiller** [comment]
See https://github.com/dotnet/runtime/issues/60422

**Comment by smourier** [comment]
> DirectX 12 Agility Sdk needs to export constants in exe, seems like this is not possible for .NET
> 
> https://devblogs.microsoft.com/directx/gettingstarted-dx12agility/#2-set-agility-sdk-parameters
> 
> > ### Method a: __declspec(dllexport) keyword
> > A can export these constants as follows using the __declspec(dllexport) keyword:
> > ```c++
> > extern "C" { __declspec(dllexport) extern const UINT D3D12SDKVersion = n;}
> > 
> > extern "C" { __declspec(dllexport) extern const char* D3D12SDKPath = u8".\\D3D12\\"; }
> > ```
> > Here n corresponds to the D3D12SDKVersion you’re using. For our first Agility SDK, this value is 4.

Answering the question about D3D12 (not the fact we cannot indeed export these from .NET currently), you can do this from .NET using [ID3D12SDKConfiguration::SetSDKVersion method](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12sdkconfiguration-setsdkversion), just interop `D3D12GetInterface`,  declare interop for `ID3D12SDKConfiguration`, and call it like this;

    conf.SetSDKVersion(611, @".\D3D12"); // with n = 611 as of latest release today

Obviously you must do that very early in your program before any "real" D3D12 call is made

**Reply by MichalPetryka** [reply]
> you can do this from .NET using [ID3D12SDKConfiguration::SetSDKVersion method](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12sdkconfiguration-setsdkversion)

The biggest issue with this is that this solution is Windows 11 only.

**Reply by smourier** [reply]
It works for Windows 10 (don't know exactly which version) and Windows 11 but only if developer mode is enabled. Do you have a better solution to this?

**Reply by MichalPetryka** [reply]
> don't know exactly which version

The documentation says `Build 20348` on the `ID3D12SDKConfiguration` interface and as per Wikipedia, the latest public build of Windows 10 is `19045`, which makes this Windows 11 only.

> Do you have a better solution to this?

As far as I know, the only options today are making a custom host in C/C++/Rust/Zig or using NativeAOT and static linking a native library that exports stuff.

**Reply by smourier** [reply]
Yes I can read too, but it does works for Windows 10 but requires developer mode enabled, try it if you don't believe me. As for custom host, it's relatively easy for comhost but currently dotnet custom host doesn't help for singlefilehost.

**Reply by weltkante** [reply]
I'm compiling my C# code as a COM library exposing a single self-defined COM interface, and wrote a small C++ app loading the library (directly, no registration required), creating the COM object, and calling into it. Saves me from writing a custom host.

**Comment by smourier** [comment]
Actually, it was such a pain that FWIW, I wrote a tool for that https://github.com/smourier/D3D12SDKVersionPatcher it will patch any exe (.NET and native ones) with D3D12 agility SDK parameters provided. IT should work for singlefilehost as well as default host.
